server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java

```diff
@@ -527,6 +531,128 @@ private void checkAndAddInEligibleTargetNode(RoutingNode targetNode) {
         }
     }
 
+    /**
+     * Triggers shard relocation for the shards that are submitted
+     * for tiering to go from local node pool to remote capable node
+     * pool.
+     */
+    public void tierShards() {
+        List<ShardRouting> shardsPendingTiering = new ArrayList<>();
+        List<RoutingNode> remoteRoutingNodes = new ArrayList<>();
+        Iterator<RoutingNode> nodes = routingNodes.mutableIterator();
+        while (nodes.hasNext()) {
+            RoutingNode node = nodes.next();
+            RoutingPool pool = RoutingPool.getNodePool(node);
+            if (pool == RoutingPool.REMOTE_CAPABLE) {
+                remoteRoutingNodes.add(node);
+            }
+        }
+
+        /* Routing Table is created after reroute and is immutable. It does not hold
+         * updated shard states with changes applied during the reroute cycle. Throughout the
+         * allocation (reroute) cycle, we need to work on RoutingNodes shard objects.
+         */
+        List<ShardRouting> indexShards = allocation.routingTable().allShards();
+        for (ShardRouting shard : indexShards) {
+            IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shard.index());
+            if (IndexModule.TieringState.HOT_TO_WARM.name()
+                .equals(indexMetadata.getSettings().get(IndexModule.INDEX_TIERING_STATE.getKey()))) {
+                List<ShardRouting> shardRoutings = allocation.routingNodes().assignedShards(shard.shardId());
+                if (shardRoutings.isEmpty()) {
+                    logger.debug(
+                        "Tiering in progress found for shards in [{}] state. Can only tier assigned shards",
+                        shard.state().toString()
+                    );
+                    continue;
+                }
+                if (shard.started()) {
+                    RoutingPool targetPool = RoutingPool.getShardPool(shard, allocation);
+                    RoutingPool currentNodePool = RoutingPool.getNodePool(allocation.routingNodes().node(shard.currentNodeId()));
+                    if (targetPool != currentNodePool) {
+                        logger.trace(
+                            "Found shard [{}] with target pool [{}] on node pool [{}]. Adding to tiering list.",
+                            shard.toString(),
+                            targetPool.toString(),
+                            currentNodePool.toString()
+                        );
+                        shardsPendingTiering.add(shard);
+                    }
+                }
+            }
+        }
+
+        if (!shardsPendingTiering.isEmpty() && remoteRoutingNodes.isEmpty()) {
+            logger.warn("No nodes available in the remote capable pool: [{}]");
+            return;
+        }
+
+        Randomness.shuffle(remoteRoutingNodes);
+        Queue<RoutingNode> nodeQueue = new ArrayDeque<>(remoteRoutingNodes);
```

Comment:
Apologies if this is explained somewhere already, but should we be trying to balance by shard count or something? If we do it randomly won't the cluster rebalance operation pick up the shard right away afterwards the tiering operation and we'll end up with duplicated shard movement?

Reply:
I was also thinking if we should assign the shards according to the order of nodes filled, from least filled to most filled. But this information is not directly available in RoutingNodes and hence adds an overhead in first calculating assigned (primary) shard count on each node. Hence, chose start node for round robin at random.
Also, allocateUnassigned for warm shards has the random assignment currently: https://github.com/neetikasinghal/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java#L65

Reply:
That makes sense -- we will have any customer rebalance logic for the warm nodes/shards then? Such as balance only primaries for example.

Reply:
that already exists here: https://github.com/neetikasinghal/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java#L203